### ISSUE TITLE
improved-base: Uncaught assertion failed: thread should have been marked

### ISSUE THREAD

**Issue by bergey** [ghcjs/ghcjs#385]
Since switching to `improved-base`, I've been getting the error above on a couple of small programs that fork threads.  Is that a bug in the runtime?  In one case I get the trace below:

```
Uncaught assertion failed: thread should have been marked
h$resolveDeadlocks @ rts.js:6506
h$gc @ rts.js:6267
h$scheduler @ rts.js:9556
h$mainLoop @ rts.js:9668
```

In the case of [my timer demo](https://github.com/bergey/gooey/tree/master/timer) I'm pretty sure that the only things I've changed since it was working yesterday are:
- GHCJS `improved-base` instead of `master` (and compiled with 7.10.2 instead of 7.8.4)
- `ghcjs-dom-0.2.1` instead of `0.1.1.3`

I'll see if I can put together a minimal example.  Let me know if I can do anything else to help debug.

**Comment by luite** [comment]
Minimal example would be great, especially if you can get rid of the deps. There have been a few more reports of differences between GHC and GHCJS, but it seems hard to reproduce the problem.

I'll still look into this some time this week, since the assertion failure may help pinpoint the cause, the other reports were just finalizers running too early.

**Comment by bergey** [comment]
Here's a small example that gives the same assertion failure as above:

``` haskell
module Main where

import           Control.Concurrent
import           Control.Concurrent.Chan
import           Control.Monad           (forever, void)

main :: IO ()
main = do
  q <- newChan
  -- loop, push to Chan
  void . forkIO . forever $ do
    writeChan q ()
    threadDelay 1000000
  -- read, read from Chan
  void . forever $ readChan q
```

**Comment by luite** [comment]
Found the problem, MVar has been changed to have a third list, `waiters`, for supporting the `readMVar` op, where threads wait for the `MVar` to be filled, without taking the value out. This list wasn't scanned, breaking an invariant for the deadlock resolution.

**Comment by luite** [comment]
should be fixed in ghcjs/shims@375e8cc41a4ceb634d3ba1ff2afe4fdd074bcb92
